crates/subspace-core-primitives/Cargo.toml

```diff
@@ -68,6 +68,7 @@ std = [
     "rand_core/std",
     "scale-info/std",
     "serde",
+    "serde_arrays",
```

Comment:
Hmm. Even in case only `serde` is enabled and `serde_arrays` is not we will encounter the same error due to feature unification.

Also while changing `cfg_attr(feature = "serde"` to `cfg_attr(feature = "std"` would solve the problem, it may be bit restrictive as you would have to enable every feature under `std` for a crate that might not require everything. 

My suggestion would be to use `dep:` syntax to stop the visibility of optional dependency as feature to users of the crate. We can create a `serde_support` feature that is composed of `dep:serde` and `dep:serde_arrays` (and maybe other pallets we do not want to expose) and include it in std.  This way optional crates are not exposed as individual feature and there will not be any configuration that can make pallet fail to compile.

Reply:
`serde_arrays` doesn't support `no_std` environment and using it in some cases, but not others will mean that sometimes serialized JSON will not be deserializable. We should keep those in sync with one another and until `serde_arrays` supports `no_std` the only way is to use `cfg_attr(feature = "std"`.

Reply:
Make sense. 

Maybe we can mark the optional dependency that we want to be internal as `dep:` so that we do not have to put blanket `cfg_attr(feature = "std"` in code and outside code can only use `std`.  I am just trying to avoid mass modification of cfg attribute if possible. :)

Reply:
It is probably not too bad of a change and certainly justified, feel free to go ahead and change what needs to be changed

Reply:
@nazar-pc I did mark additional dependency as `dep:` so that crate users cannot use them independently from outside. Unfortunately, marking them `dep:` means the feature is hidden at the crate level as well so I did change from `serde` and `thiserror` to `std`.

Does that work for you?

Reply:
I really don't see why it was necessary. `dep:` is useful when you want to have feature and dependency with the same name. Changing it just so you have to change even more code doesn't quite make sense to me. In fact you could have changed source code and made sure that both `serde` and `serde_arrays` are either both enabled or both not (and add `hex/serde` feature there too rather than to `std`). For `thiserror` and `rand` I see no error to change them.

Reply:
@nazar-pc The benefit of `dep:` is now there are no separate `serde`, `serde_arrays` and `thiserror` features. They are all integral part of `std` which would eliminate chance of misconfiguring the pallet. This also indirectly simplifies codebase as instead of checking for every feature related to `serde` we can just check for `std`.

Reply:
It is also a drawback, the reason we didn't have `std` in code in those places is so that you can enable `serde` without enabling the rest of the `std`. Similarly fr `thiserror`, but less useful.

It makes code more intentional and better self-documented.

Reply:
And `hex/serde` feature too, we use it for some fields

Reply:
@nazar-pc Done. used `cfg-if` macro to check if either all are enabled or none are.